### PR TITLE
Add a visual hint when the user has exceeded the rate limit

### LINKED ISSUE #56455: JetBrains: add visual hint when the user has exceeded the rate limit
Steps to reproduce:

- Exceed the rate limit somehow
- Send a chat message or trigger autocomplete
- Currently, nothing happens, Cody just silently stops working
- Ideally, Cody should inform me that the rate limit has been exceeded and educates me about how to fix the problem. I had to look at the IntelliJ logs to see a message about `Completion request failed: {"message": "rate limit exceeded"}`

### PR DESCRIPTION
**Pull request by mkondratek**
Cody part: https://github.com/sourcegraph/cody/pull/1662.

Fixes: https://github.com/sourcegraph/sourcegraph/issues/56455.

## Test plan
1. mock Cody responses to throw the error
2. `runIde` with a local clone of Cody
3. play with autocompletion/chat

### CODE REVIEW

File: src/main/java/com/sourcegraph/cody/vscode/InlineAutocompleteItemProvider.java [hunk 1]
```diff
@@ -1,11 +0,0 @@
-package com.sourcegraph.cody.vscode;
-
-import java.util.concurrent.CompletableFuture;
-
-public abstract class InlineAutocompleteItemProvider {
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/vscode/InlineAutocompleteItemProvider.java:5]: unrel, unused

**Comment by mkondratek** [comment]
<details>
<summary>Current state demo</summary>

## Autocomplete (Invoke)

https://github.com/sourcegraph/jetbrains/assets/19799111/cf9f15ff-0777-495e-8a96-0edaf81faf52

## Chat

https://github.com/sourcegraph/jetbrains/assets/19799111/0446c3e7-04f6-4b48-a052-e2cd02f27b81




</details>

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.sourcegraph.cody.agent.protocol
+
+import com.sourcegraph.cody.vscode.InlineAutocompleteItem
+
+class InlineAutocompleteList {
+  val items: List<InlineAutocompleteItem>? = null
```
Comment (Strum355) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt]: This can be non-null, according to agent-protocol.ts

**Review by Strum355 — APPROVED** [review]

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt (hunk 2 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt]: so `emptyList()` default is ok, right?

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/RateLimitError.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          json: JsonElement,
+          typeOfT: Type?,
+          context: JsonDeserializationContext?
+      ): RateLimitError {
+        val jsonObject = json.asJsonObject
+        val errorObject = jsonObject.getAsJsonObject("error")
+
+        val limit = errorObject.getAsJsonPrimitive("limit")?.asInt
+        val retryAfter =
+            errorObject.getAsJsonPrimitive("retryAfter")?.asString?.let { parseOffsetDateTime(it) }
```
Comment (Strum355) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/RateLimitError.kt]: I think these are equivalent? But I might be wrong, so lmk if thats the case :slightly_smiling_face: 
```suggestion
        val jsonObject = json.asJsonObject
        val errorObject = jsonObject["error"].asJsonObject

        val limit = errorObject["limit"]?.asInt
        val retryAfter =
            errorObject["retryAfter"]?.asString?.let { parseOffsetDateTime(it) }
```

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt (hunk 2 above)
Comment (Strum355) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/InlineAutocompleteList.kt]: Yup thats correct

**Pull request merged**